### PR TITLE
Notification Service

### PR DESCRIPTION
**Pull request by SandroGlueck**
## Implemented special service for notifying users in certain cases (item is closed, item is rejected, ...)

* Added new lambda service with two handlers (item_closed, item_rejected)
* Added sender implementations for notifications via mail and telegram
* Added base class for further senders
* Added tests

## Extended existing lambda functions to trigger notifications

* Extended review_service.update_item
* Extended admin_service.update_item

## Extended lambda core layer

* Added class to provide Boto3 client
* Added class to publish EventBridge events
* Added classes for common response patterns
* Added handler to retrieve message templates from S3
* tests

## Misc

* removed old notifications.py
* adapted build specs

### CODE REVIEW

File: lambda_functions/admin_service/update_item.py [hunk 1]
```diff
@@ -68,6 +69,11 @@ def update_item(event, context, is_test=False, session=None):
             return response_cors
 
     item = update_object(item, is_test, session)
+
+    if(item.status == 'rejected'):
+        EventPublisher().publish_event(
```
Comment (ChristianBay) [inline lambda_functions/admin_service/update_item.py:74]: Ich kenne jetzt Python zu wenig. Aber sollte der EventPublisher nicht eher eine singleton Klasse sein. 
Also das alle denselben EventPublisher verwenden.

Oder alternativ eine static function.

Aber wie gesagt, ich bin nicht tief genug drin, eventuell ist es auch Blödsinn was ich hier schreibe ;-)

File: lambda_functions/notification_service/src/item_closed_handler.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        rating = round(item.result_score, 1)
+        rating_text = get_rating_text(rating)
+
+        parameters = dict(
+            rating=rating, rating_text=rating_text, item_id=item.id, content=item.content
+        )
+
+        for submission in item.submissions:
+            if submission.mail is not None:
+                mail_sender.send_notification(
```
Comment (ChristianBay) [inline lambda_functions/notification_service/src/item_closed_handler.py:52]: Da ja bei jeder send_notificiation eine Exception fliegen kann, sollte man meiner Meinung das immer mit einem try-catch umschließen, damit die anderen Submissions/Handler trotzdem ausgeführt werden können.

File: lambda_functions/admin_service/update_item.py (hunk 1 above)
Comment (SandroGlueck) [reply lambda_functions/admin_service/update_item.py:74]: Sollte in dem Fall glaube ich tatsächlich relativ egal sein.
Der Code wird in der Lambda Funktion ohnehin nur einmal durchlaufen.

Hatte das ganze nur in eine Klasse verpackt, falls tatsächlich mal der Fall eintritt, dass in einer Lambda mehrere Events gepublished werden müssen.

File: lambda_functions/notification_service/src/item_closed_handler.py (hunk 2 above)
Comment (SandroGlueck) [reply lambda_functions/notification_service/src/item_closed_handler.py:52]: Hast absolut recht, werde ich beheben

File: lambda_layers/core_layer/python/core_layer/test/test_responses.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+    monkeypatch.setenv("CORS_ALLOW_ORIGIN", ORIGIN)
+
+    response = NoContent({})
+
+    assert response.statusCode == 204
+    assert response.body is None
+
+
+def test_bad_request(monkeypatch):
```
Comment (ChristianBay) [inline lambda_layers/core_layer/python/core_layer/test/test_responses.py:166]: **test_bad_request** function name is wrong

File: lambda_layers/core_layer/python/core_layer/responses.py [hunk 4]
```diff
@@ -10,21 +42,30 @@ def __init__(self, message: str = None, exception: Exception = None):
...
+            return
+
+        self.body = {}
+        self.body['message'] = message
+        self.body['exception'] = exception
 
 
-class BadRequest:
-    def __init__(self, message: str = None):
+class BadRequest(ResponseBase):
```
Comment (ChristianBay) [inline lambda_layers/core_layer/python/core_layer/responses.py:54]: Very nice implementation

**Review by ChristianBay — APPROVED** [review]

File: lambda_functions/notification_service/src/item_closed_handler.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            if item is None:
+                return BadRequest(event,
+                                  f"No item was found with the given item_id [{item_id}].")
+
+            # TODO: This implementation is not ideal: 1.55 is rounded to 1.5. However, 1.56 is correctly rounded to 1.6.
+            rating = round(item.result_score, 1)
+            rating_text = get_rating_text(rating)
+
+            parameters = dict(
+                rating=rating, rating_text=rating_text, item_id=item.id, content=item.content
```
Comment (PotSchmoell) [inline lambda_functions/notification_service/src/item_closed_handler.py:41]: I don't really understand where the actual message to send comes from. Shouldn't this line handover an `html` parameter to be replaced in the sender? 
Or am I misunderstanding something? 
Also I failed to find the line/file where the actual text/html, which is being sent is defined.

Comment (PotSchmoell) [reply lambda_functions/notification_service/src/item_closed_handler.py:41]: Nevemind @SandroGlueck, I got it know.

**Comment by PotSchmoell** [comment]
Are you sure, that we really want to have the message templates in S3 instead of including them in the repo? 
Imho this has the following downsides:
- It makes the code harder to understand because not all the information is in the repository
- In the current state, we have no staging for the template files
- Everybody who wants to adapt the  template needs access to AWS and S3
- It inflates our buildspec and template files
What do you think of the alternative of putting the templates into an `assets` folder in the notification service?

**Comment by SandroGlueck** [comment]
> Are you sure, that we really want to have the message templates in S3 instead of including them in the repo?

I considered putting the files in the repo or storing it in the DB.
If I remember correctly, the two of us had then discussed the possibility of storing it in an S3 bucket.

> Imho this has the following downsides:
> 
> * It makes the code harder to understand because not all the information is in the repository

I can further clarify that by some comments, if that helps.

> * In the current state, we have no staging for the template files

We could easily add multiple buckets for each stage or add folders in that bucket for the different stages - for now, I only added one to keep things simple.

> * Everybody who wants to adapt the  template needs access to AWS and S3

If we put it in the repo, everybody would need access rights to the repo and we would need to deploy a new version every time the templates change.
We could extend the admin service to view/change/update notification templates.

> * It inflates our buildspec and template files

It only adds a variable to one of the template files...

>   What do you think of the alternative of putting the templates into an `assets` folder in the notification service?

I personally feel that the options S3 and DB are better suited than having them in the repo (S3 slightly easier to manage changes).
However, we can discuss this topic with the others, if you want.

**Comment by PotSchmoell** [comment]
Alright, you convinced me :)

**Comment by SandroGlueck** [comment]
That was easy 😅

**Pull request merged**